tests/test_signals.py

```diff
@@ -21,11 +19,22 @@ def test_churn(self):
         assert np.isclose(churn(self.up, self.up_down), 1)
```

Comment:
Not helpful for the failures you're seeing, but just a couple of comments:

1. Check out the [numpy.testing](https://numpy.org/doc/stable/reference/routines.testing.html) module that provides assertion methods that - especially when you're using the built-in `unittest` - are a bit friendlier than naked `assert`s:

```
>>> import numpy.testing as t

>>> t.assert_allclose(churn(self.up, self.up_down), 1.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jelford/src/numerai/numerai-tools/.venv/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 1504, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/jelford/src/numerai/numerai-tools/.venv/lib/python3.12/site-packages/numpy/testing/_private/utils.py", line 797, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 1 / 1 (100%)
Max absolute difference: 0.1
Max relative difference: 0.09090909
 x: array(1)
 y: array(1.1)

```

What do you buy here?
- automatic comparison rather than this awkward `assert result == expected, result` setup where you have to pass your own failure message (not always bad, but for simple numerical comparisons...)
- you also get the same `atol`, `rtol` args that you already have on `np.isclose` so not a big change but still good to know about.

This is an extension of the same reasoning why `unittest.TestCase` has a bunch of `assertX` methods that might seem a bit redundant: just better failure messages.

I'd really recommend using naked `assert` only if running under `pytest` which injects a  bunch of magic to yield more helpful error messages (in the early days that was one of its big features for pytest).


2. Just a style thing - I often like to write `result` for the result of the unit under test - to make really explicit at a glance what's under test. It's pretty clear in these very simple functional tests, but it's a nice habit when you have more setup and more complicated assertions. Certainly more descriptive than `tmp`. I'll just mention too: if I give a name to the expectation, I normally call it `expected` for similar reasons (again, most useful if constructing your expected output is non-trivial and you can't just write it in line as as a single number in the assertion - I find what you've done here of just writing the expectation in the assertion clearer than introducing a new line for no reason).

3. Again just a hint for if you do want to look at using `pytest`: the list of scenarios you've got in your tests here are like a perfect textbook example of when [pytest.mark.parameterize](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html) can help out; basically you define the test once and then just list all the scenarios. I think the docs linked there sort of jump too quickly into the "advanced" features without motivating much, so I'll just link [an example from over in trading](https://github.com/Numerai-CM/monorepo/blob/40b168ef29f88833ea72555a39e2a5d9cf3a2ebd/trading/trading-pipeline/tests/trading_pipeline/oems_portal/trading_calendar/test_trading_schedule_calculation.py#L110) where we have  a list of ~80 examples reflecting the evolution of trading schedules over time, and it's easy to add a quick "what if" if we've got a holiday coming up or such. This example really profits from the scenario-based test specification, because we have a bit of boilerplate mocking stuff to set up in each test, so it would be a lot of fragile copypasta to repeat as separate functions.